src/lib/encoder/Decoder.h

```diff
@@ -64,14 +64,28 @@ inline unsigned Decoder::do_decode(CimbReader& reader, STREAM& ostream)
 {
 	bitbuffer bb(_bitsPerOp * 1550);
 	std::vector<unsigned> interleaveLookup = Interleave::interleave_reverse(reader.num_reads(), _interleaveBlocks, _interleavePartitions);
+	std::array<PositionData, 12400> colorPositions; // 12400 = 1500 * 8, aka the number of cells. This should probably be determined at runtime
+
+	// read symbols first
 	while (!reader.done())
 	{
-		// reader should probably be in charge of the cell index (i) calculation
+		// reader is in charge of the cell index (i) calculation
 		// we can compute the bitindex ('index') here, but only the reader will know the right cell index...
-		unsigned bits = 0;
-		unsigned i = reader.read(bits);
-		unsigned bitPos = interleaveLookup[i] * _bitsPerOp;
+		PositionData pos;
+		unsigned bits = reader.read(pos);
+
+		unsigned bitPos = interleaveLookup[pos.i] * _bitsPerOp;
 		bb.write(bits, bitPos, _bitsPerOp);
+
+		colorPositions[pos.i] = {bitPos, pos.x, pos.y};
+	}
+
+	// then decode colors.
+	// the symbol+color decode could be done as one pass, but doing it as two gives us more control of the caches
+	for (const PositionData& p : colorPositions)
+	{
+		unsigned bits = reader.read_color(p);
```

Comment:
Later, we read the color. We *could* sort `colorPositions` before this second loop (and I tried a few sorts), but in practice sorting by index (what the first loop does) seems to be pretty good.

Sorting by `x` is a performance disaster, which seems like more validation of the theory.